src/term/transform/extract_matches.rs

```diff
@@ -0,0 +1,215 @@
+          write!(f, "\nConsider using a match block instead")?;
+        }
+
+        Ok(())
+      }
+    }
+  }
+}
+
+impl Term {
+  pub fn check_matches<'book>(
+    pats: &[Pattern],
+    adts: &'book BTreeMap<Name, Adt>,
+    ctrs: &'book HashMap<Name, Name>,
+  ) -> Result<&'book Adt, MatchError> {
+    let ty = infer_arg_type(pats.iter(), ctrs).map_err(MatchError::Infer)?;
+
+    let Type::Adt(nam) = ty else { unreachable!() };
+
+    let Adt { ctrs } = &adts[&nam];
+
+    let mut missing: HashSet<_> = ctrs.keys().cloned().collect();
+
+    for rule in pats {
+      let Pattern::Ctr(nam, args) = rule else { unreachable!() };
+
+      let mut binds = HashSet::new();
+
+      for arg in args {
+        for bind in arg.names() {
+          if !binds.insert(bind) {
+            return Err(MatchError::Repeated(bind.clone()));
+          }
+        }
+      }
+
+      missing.remove(nam);
+    }
+
+    if !missing.is_empty() {
+      return Err(MatchError::Missing(missing));
+    }
+
+    Ok(&adts[&nam])
+  }
+
+  fn extract_matches(
+    &mut self,
+    def_name: &Name,
+    book: &mut MatchesBook,
+    match_count: &mut usize,
+  ) -> Result<(), MatchError> {
+    match self {
+      Term::Match { scrutinee: box Term::Var { nam }, arms } => {
+        if arms.is_empty() {
+          return Err(MatchError::Empty);
+        }
+
+        for (_, term) in arms.iter_mut() {
+          term.extract_matches(def_name, book, match_count)?;
+        }
+
+        *match_count += 1;
+
+        let arms = std::mem::take(arms);
+        let nam = std::mem::take(nam);
+
+        *self = match_to_def(nam, &arms, def_name, book, match_count)?;
+      }
+
+      Term::Lam { bod, .. } | Term::Chn { bod, .. } => {
+        bod.extract_matches(def_name, book, match_count)?;
+      }
+      Term::App { fun: fst, arg: snd, .. }
+      | Term::Let { pat: Pattern::Var(_), val: fst, nxt: snd }
+      | Term::Dup { val: fst, nxt: snd, .. }
+      | Term::Tup { fst, snd }
+      | Term::Sup { fst, snd, .. }
+      | Term::Opx { fst, snd, .. } => {
+        fst.extract_matches(def_name, book, match_count)?;
+        snd.extract_matches(def_name, book, match_count)?;
+      }
+      Term::Var { .. } | Term::Lnk { .. } | Term::Num { .. } | Term::Ref { .. } | Term::Era => {}
+
+      Term::Match { .. } => unreachable!("Scrutinee of match expression should have been extracted already"),
+      Term::Let { pat, .. } => {
+        unreachable!("Destructor let expression should have been desugared already. {pat}")
+      }
+    }
+
+    Ok(())
+  }
+}
+
+/// Transforms a match into a new definition with every arm of `arms` as a rule.
+/// The result is the new def applied to the scrutinee followed by the free vars of the arms.
+fn match_to_def(
+  scrutinee: Name,
+  arms: &[(Pattern, Term)],
+  def_name: &Name,
+  book: &mut MatchesBook,
+  match_count: &mut usize,
+) -> Result<Term, MatchError> {
+  let current_count = *match_count;
```

Comment:
We can remove this now, the function does not need to receive a `&mut usize` anymore but was left here in an earlier refactor.

Reply:
The function does not need to return a result anymore too